File: src/test/java/org/mapdb/DBMakerTest.java
package org.mapdb;

import org.junit.Test;

import java.io.File;
import java.io.IOException;
import java.io.Serializable;
import java.lang.reflect.Field;
import java.util.*;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.ConcurrentNavigableMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.locks.LockSupport;

import static org.junit.Assert.*;

@SuppressWarnings({ "unchecked", "rawtypes" })
public class DBMakerTest{

    
    private void verifyDB(DB db) {
        Map m = db.hashMap("test");
        m.put(1,2);
        assertEquals(2, m.get(1));
    }


    @Test
    public void testNewMemoryDB() throws Exception {
        DB db = DBMaker
                .memoryDB()
                .transactionDisable()
                .make();
        verifyDB(db);
    }


    @Test
    public void testNewFileDB() throws Exception {
        File f = TT.tempDbFile();
        DB db = DBMaker.fileDB(f)
                .transactionDisable().make();
        verifyDB(db);
    }

    @Test
    public void testDisableTransactions() throws Exception {
        DBMaker.memoryDB().make();
    }

    @Test
    public void testDisableCache() throws Exception {
        DB db = DBMaker
                .memoryDB()
                .transactionDisable()
                 .make();
        verifyDB(db);
        Store s = Store.forDB(db);
        assertEquals(s.getClass(), StoreDirect.class);
    }



    @Test
    public void testMake() throws Exception {
        DB db = DBMaker
                .fileDB(TT.tempDbFile())
                .transactionDisable()
                .make();
        verifyDB(db);
        //check default values are set
        Engine w =  db.engine;
        Store store = Store.forDB(db);
        assertNull(store.caches);
        StoreDirect s = (StoreDirect) store;
        assertTrue(s.vol instanceof Volume.RandomAccessFileVol);
    }

    @Test
    public void testCacheHashTableEnable() throws Exception {
        DB db = DBMaker
                .fileDB(TT.tempDbFile())
                .cacheHashTableEnable()
                .transactionDisable()
                .make();
        verifyDB(db);
        //check default values are set
        Engine w =  db.engine;
        Store store = Store.forDB(db);
        assertTrue(store.caches[0] instanceof Store.Cache.HashTable);
        assertEquals(1024 * 2, ((Store.Cache.HashTable) store.caches[0]).items.length * store.caches.length);
        StoreDirect s = (StoreDirect) store;
        assertTrue(s.vol instanceof Volume.RandomAccessFileVol);
    }

    @Test
    public void testMakeMapped() throws Exception {
        DB db = DBMaker
                .fileDB(TT.tempDbFile())
                .transactionDisable()
                .mmapFileEnable()
                .make();
        verifyDB(db);
        //check default values are set
        Engine w = db.engine;
        Store store = Store.forDB(db);
        StoreDirect s = (StoreDirect) store;
        assertTrue(s.vol instanceof Volume.MappedFileVol);
    }

    @Test
    public void testCacheHardRefEnable() throws Exception {
        DB db = DBMaker
                .memoryDB()
                .transactionDisable()
                .cacheHardRefEnable()
                .make();
        verifyDB(db);
        Store store = Store.forDB(db);
        assertTrue(store.caches[0].getClass() == Store.Cache.HardRef.class);
    }

    @Test
    public void testCacheWeakRefEnable() throws Exception {
        DB db = DBMaker
                .memoryDB()
                .transactionDisable()
                .cacheWeakRefEnable()
                .make();
        verifyDB(db);
        Store store = Store.forDB(db);
        Store.Cache cache = store.caches[0];
        assertTrue(cache.getClass() == Store.Cache.WeakSoftRef.class);
        assertTrue(((Store.Cache.WeakSoftRef) cache).useWeakRef);
    }


    @Test
    public void testCacheSoftRefEnable() throws Exception {
        DB db = DBMaker
                .memoryDB()
                .transactionDisable()
                .cacheSoftRefEnable()
                .make();
        verifyDB(db);
        Store store = Store.forDB(db);
        assertTrue(store.caches[0].getClass() == Store.Cache.WeakSoftRef.class);
        assertFalse(((Store.Cache.WeakSoftRef) store.caches[0]).useWeakRef);
    }

    @Test
    public void testCacheLRUEnable() throws Exception {
        DB db = DBMaker
                .memoryDB()
                .transactionDisable()
                .cacheLRUEnable()
                .make();
        verifyDB(db);
        Store store = Store.forDB(db);
        assertTrue(store.caches[0].getClass() == Store.Cache.LRU.class);
        db.close();
    }

    @Test
    public void testCacheSize() throws Exception {
        DB db = DBMaker
                .memoryDB()
                .transactionDisable()
                .cacheHashTableEnable()
                .cacheSize(1000)
                .make();
        verifyDB(db);
        Store store = Store.forDB(db);
        assertEquals(1024, ((Store.Cache.HashTable) store.caches[0]).items.length * store.caches.length);
    }


    @Test public void read_only() throws IOException {
        File f = TT.tempDbFile();
        DB db = DBMaker.fileDB(f).make();
        db.close();
        db = DBMaker
                .fileDB(f)
                .deleteFilesAfterClose()
                .readOnly()
                .make();
        assertTrue(db.engine instanceof Engine.ReadOnlyWrapper);
        db.close();
    }


    @Test(expected = DBException.WrongConfig.class)
    public void reopen_wrong_checksum() throws IOException {
        File f = TT.tempDbFile();
        DB db = DBMaker.fileDB(f).make();
        db.close();
        db = DBMaker
                .fileDB(f)
                .deleteFilesAfterClose()

                .checksumEnable()
                .make();
        Engine w = db.engine;
        assertTrue(w instanceof TxEngine);

        Store s = Store.forEngine(w);
        assertTrue(s.checksum);
        assertTrue(!s.compress);
        assertTrue(!s.encrypt);
        db.close();
    }


    @Test public void checksum() throws IOException {
        File f = TT.tempDbFile();
        DB db = DBMaker
                .fileDB(f)
                .deleteFilesAfterClose()
                .checksumEnable()
                .make();

        Store s = Store.forDB(db);
        assertTrue(s.checksum);
        assertTrue(!s.compress);
        assertTrue(!s.encrypt);
        db.close();
    }

    @Test public void encrypt() throws IOException {
        File f = TT.tempDbFile();
        DB db = DBMaker
                .fileDB(f)
                .deleteFilesAfterClose()
                .encryptionEnable("adqdqwd")
                .make();
        Store s = Store.forDB(db);
        assertTrue(!s.checksum);
        assertTrue(!s.compress);
        assertTrue(s.encrypt);
        db.close();
    }


    @Test(expected = DBException.WrongConfig.class)
    public void reopen_wrong_encrypt() throws IOException {
        File f = TT.tempDbFile();
        DB db = DBMaker.fileDB(f).make();
        db.close();
        db = DBMaker
                .fileDB(f)
                .deleteFilesAfterClose()
                .encryptionEnable("adqdqwd")
                .make();
        Store s = Store.forDB(db);
        assertTrue(!s.checksum);
        assertTrue(!s.compress);
        assertTrue(s.encrypt);
        db.close();
    }


    @Test public void compress() throws IOException {
        File f = TT.tempDbFile();
        DB db = DBMaker
                .fileDB(f)
                .deleteFilesAfterClose()
                .compressionEnable()
                .make();
        Store s = Store.forDB(db);
        assertTrue(!s.checksum);
        assertTrue(s.compress);
        assertTrue(!s.encrypt);
        db.close();
    }

    @Test(expected = DBException.WrongConfig.class)
    public void reopen_wrong_compress() throws IOException {
        File f = TT.tempDbFile();
        DB db = DBMaker.fileDB(f).make();
        db.close();
        db = DBMaker
                .fileDB(f)
                .deleteFilesAfterClose()
                .compressionEnable()
                .make();
        Engine w = db.engine;
        assertTrue(w instanceof TxEngine);
        Store s = Store.forEngine(w);
        assertTrue(!s.checksum);
        assertTrue(s.compress);
        assertTrue(!s.encrypt);

        db.close();
    }



    @Test public void close_on_jvm_shutdown(){
        DBMaker
                .tempFileDB()
                .closeOnJvmShutdown()
                .deleteFilesAfterClose()
                .make();
    }

    @Test public void tempTreeMap(){
        ConcurrentNavigableMap<Long,String> m = DBMaker.tempTreeMap();
        m.put(111L,"wfjie");
        assertTrue(m.getClass().getName().contains("BTreeMap"));
    }

    @Test public void tempHashMap(){
        ConcurrentMap<Long,String> m = DBMaker.tempHashMap();
        m.put(111L, "wfjie");
        assertTrue(m.getClass().getName().contains("HTreeMap"));
    }

    @Test public void tempHashSet(){
        Set<Long> m = DBMaker.tempHashSet();
        m.add(111L);
        assertTrue(m.getClass().getName().contains("HTreeMap"));
    }

    @Test public void tempTreeSet(){
        NavigableSet<Long> m = DBMaker.tempTreeSet();
        m.add(111L);
        assertTrue(m.getClass().getName().contains("BTreeMap"));
    }



    @Test public void keys_value_matches() throws IllegalAccessException {
        Class c = DBMaker.Keys.class;
        Set<Integer> s = new TreeSet<Integer>();
        for (Field f : c.getDeclaredFields()) {
            f.setAccessible(true);
            String value = (String) f.get(null);

            String expected = f.getName().replaceFirst("^[^_]+_","");
            assertEquals(expected, value);
        }
    }

    File folderDoesNotExist = new File("folder-does-not-exit/db.aaa");

    @Test(expected = DBException.VolumeIOError.class)
    public void nonExistingFolder(){
        DBMaker.fileDB(folderDoesNotExist).make();
    }

    @Test(expected = DBException.VolumeIOError.class)
    public void nonExistingFolder3(){
        DBMaker.fileDB(folderDoesNotExist).mmapFileEnable().make();
    }


    @Test(expected = DBException.VolumeIOError.class)
    public void nonExistingFolder2(){
        DBMaker
                .fileDB(folderDoesNotExist)
                .snapshotEnable()
                .commitFileSyncDisable()
                .makeTxMaker();
    }

    @Test public void treeset_pump_presert(){
        List unsorted = Arrays.asList(4,7,5,12,9,10,11,0);

        NavigableSet<Integer> s = DBMaker.memoryDB().transactionDisable().make()
                .treeSetCreate("t")
                .pumpPresort(10)
                .pumpSource(unsorted.iterator())
                .make();

        assertEquals(Integer.valueOf(0),s.first());
        assertEquals(Integer.valueOf(12), s.last());
    }

    @Test public void treemap_pump_presert(){
        List unsorted = Arrays.asList(4,7,5,12,9,10,11,0);

        BTreeMap<Integer,Integer> s = DBMaker.memoryDB().transactionDisable().make()
                .treeMapCreate("t")
                .pumpPresort(10)
                .pumpSource(unsorted.iterator(), Fun.extractNoTransform())
                .make();

        assertEquals(Integer.valueOf(0),s.firstEntry().getKey());
        assertEquals(Integer.valueOf(12), s.lastEntry().getKey());
        s.close();
    }

    @Test public void heap_store(){
        DB db = DBMaker.heapDB().make();
        Engine  s = Store.forDB(db);

        assertTrue(s instanceof StoreHeap);
        db.close();
    }

    @Test public void executor() throws InterruptedException {
        if(TT.scale()==0)
            return;
        final DB db = DBMaker.heapDB().executorEnable().make();
        assertNotNull(db.executor);
        assertFalse(db.executor.isTerminated());

        final AtomicBoolean b = new AtomicBoolean(true);

        Runnable r = new Runnable() {
            @Override
            public void run() {
                while(b.get()) {
                    LockSupport.parkNanos(10);
                }
            }
        };

        db.executor.execute(r);

        final AtomicBoolean closed = new AtomicBoolean();
        new Thread(){
            @Override
            public void run() {
                db.close();
                closed.set(true);
            }
        }.start();

        Thread.sleep(1000);
        assertTrue(db.executor.isShutdown());

        //shutdown the task
        b.set(false);
        Thread.sleep(2000);
        assertTrue(closed.get());
        assertNull(db.executor);
        db.close();
    }

    @Test public void temp_HashMap_standalone(){
        HTreeMap m = DBMaker.tempHashMap();
        assertTrue(m.closeEngine);
        m.close();
    }

    @Test public void temp_TreeMap_standalone(){
        BTreeMap m = DBMaker.tempTreeMap();
        assertTrue(m.closeEngine);
        m.close();
    }

    @Test public void temp_HashSet_standalone() throws IOException {
        HTreeMap.KeySet m = (HTreeMap.KeySet) DBMaker.tempHashSet();
        assertTrue(m.getHTreeMap().closeEngine);
        m.close();
    }

    @Test public void temp_TreeSet_standalone() throws IOException {
        BTreeMap.KeySet m = (BTreeMap.KeySet) DBMaker.tempTreeSet();
        assertTrue(((BTreeMap)m.m).closeEngine);
        m.close();
    }


    @Test public void metricsLog(){
        ScheduledExecutorService s = Executors.newSingleThreadScheduledExecutor();
        DB db = DBMaker.memoryDB()
                .metricsEnable(11111)
                .metricsExecutorEnable(s)
                .make();

        //TODO test task was scheduled with correct interval
        assertTrue(s==db.metricsExecutor);
        assertNull(db.executor);
        db.close();
    }

    @Test public void storeExecutor(){
        ScheduledExecutorService s = Executors.newSingleThreadScheduledExecutor();
        DB db = DBMaker.memoryDB()
                .storeExecutorPeriod(11111)
                .storeExecutorEnable(s)
                .make();

        //TODO test task was scheduled with correct interval
        assertTrue(s==db.storeExecutor);
        assertNull(db.executor);
        db.close();
    }


    @Test public void cacheExecutor(){
        ScheduledExecutorService s = Executors.newSingleThreadScheduledExecutor();
        DB db = DBMaker.memoryDB()
                .cacheExecutorPeriod(11111)
                .cacheExecutorEnable(s)
                .make();

        //TODO test task was scheduled with correct interval
        assertTrue(s==db.cacheExecutor);
        assertNull(db.executor);
        db.close();
    }


    @Test public void asyncWriteCache(){
        DB db = DBMaker.memoryDB()
                .asyncWriteEnable()
                .transactionDisable()
                .make();
        assertEquals(StoreCached.class, Store.forDB(db).getClass());
        db.close();
    }

    @Test public void asyncWriteQueueSize(){
        DB db = DBMaker.memoryDB()
                .asyncWriteEnable()
                .asyncWriteQueueSize(12345)
                .transactionDisable()
                .make();
        StoreCached c = (StoreCached) Store.forDB(db);
        assertEquals(12345,c.writeQueueSize);
        db.close();
    }


    @Test public void hashmap_segmented(){
        HTreeMap m = DBMaker
                .hashMapSegmentedMemory()
                .make();

        if(HTreeMap.SEG==1)
            return;

        assertNotSame(m.engines[0], m.engines[1]);

        StoreDirect s = (StoreDirect) m.engines[0];
        assertSame(Store.NOLOCK, s.locks[0].readLock());
        assertSame(Store.NOLOCK, s.locks[0].writeLock());
        assertEquals(1, s.locks.length);
        assertFalse(s.isClosed());

        m.close();

        for(Engine e:m.engines){
            assertTrue(e.isClosed());
        }
    }

    @Test public void hashmap_segmented_expiration(){
        HTreeMap m = DBMaker
                .hashMapSegmentedMemory()
                .expireAfterWrite(100)
                .executorEnable()
                .make();

        if(HTreeMap.SEG==1)
            return;

        assertNotSame(m.engines[0], m.engines[1]);

        StoreDirect s = (StoreDirect) m.engines[0];
        assertSame(Store.NOLOCK, s.locks[0].readLock());
        assertSame(Store.NOLOCK, s.locks[0].writeLock());
        assertEquals(1, s.locks.length);
        assertFalse(s.isClosed());

        m.close();
        assertTrue(m.executor.isTerminated());

        for(Engine e:m.engines){
            assertTrue(e.isClosed());
        }
    }

    @Test public void fileChannel(){
        DB db = DBMaker.fileDB(TT.tempDbFile())
                .fileChannelEnable()
                .transactionDisable().make();
        StoreDirect d = (StoreDirect) Store.forDB(db);
        assertEquals(Volume.FileChannelVol.class, d.vol.getClass());
        db.close();
    }


    @Test public void fileMmapCleanerHack_file(){
        DB db = DBMaker.fileDB(TT.tempDbFile())
                .fileMmapEnable()
                .transactionDisable()
                .make();
        assertFalse(((Volume.MappedFileVol)((StoreDirect) db.engine).vol).cleanerHackEnabled);
        db.close();

        db = DBMaker.fileDB(TT.tempDbFile())
                .fileMmapEnable()
                .fileMmapCleanerHackEnable()
                .transactionDisable()
                .make();
        assertTrue(((Volume.MappedFileVol) ((StoreDirect) db.engine).vol).cleanerHackEnabled);
        db.close();
    }


    @Test public void fileMmapCleanerHack_memory(){
        DB db = DBMaker.memoryDirectDB()
                .transactionDisable()
                .make();
        assertFalse(((Volume.ByteBufferVol) ((StoreDirect) db.engine).vol).cleanerHackEnabled);
        db.close();

        db = DBMaker.memoryDirectDB()
                .fileMmapCleanerHackEnable()
                .transactionDisable()
                .make();
        assertTrue(((Volume.ByteBufferVol) ((StoreDirect) db.engine).vol).cleanerHackEnabled);
        db.close();
    }

    @Test public void file_locked() throws IOException {
        File f = File.createTempFile("mapdbTest", "mapdb");
        DB db = DBMaker.fileDB(f).transactionDisable().make();

        StoreDirect s = (StoreDirect) db.getEngine();
        assertTrue(s.vol.getFileLocked());
        assertNull(s.fileLockHeartbeat);
        db.close();
    }


    @Test public void file_locked_disabled() throws IOException {
        File f = File.createTempFile("mapdbTest","mapdb");
        DB db = DBMaker.fileDB(f).transactionDisable()
                .fileLockDisable()
                .make();

        StoreDirect s = (StoreDirect) db.getEngine();
        assertFalse(s.vol.getFileLocked());
        assertNull(s.fileLockHeartbeat);
        db.close();
    }


    @Test public void file_locked_disabled_wal() throws IOException {
        File f = File.createTempFile("mapdbTest","mapdb");
        DB db = DBMaker.fileDB(f)
                .fileLockDisable()
                .make();

        StoreWAL s = (StoreWAL) db.getEngine();
        assertFalse(s.vol.getFileLocked());
        //TODO check WAL size increment
//        assertFalse(s.wal.curVol.getFileLocked());
        assertNull(s.fileLockHeartbeat);
        db.close();
    }


    @Test public void file_locked_disabled_append() throws IOException {
        File f = File.createTempFile("mapdbTest","mapdb");
        DB db = DBMaker.appendFileDB(f)
                .fileLockDisable()
                .make();

        StoreAppend s = (StoreAppend) db.getEngine();
        assertFalse(s.wal.curVol.getFileLocked());
        assertNull(s.fileLockHeartbeat);
        db.close();
    }

    @Test public void file_locked_heartbeat() throws IOException {
        File f = File.createTempFile("mapdbTest","mapdb");
        DB db = DBMaker.fileDB(f).transactionDisable()
                .fileLockHeartbeatEnable()
                .make();

        StoreDirect s = (StoreDirect) db.getEngine();
        assertFalse(s.vol.getFileLocked());

        assertTrue(s.fileLockHeartbeat.isLocked());
        assertEquals(new File(f.getPath() + ".lock"), s.fileLockHeartbeat.getFile());
        db.close();
    }

    @Test public void allocate_start_size(){
        DB db = DBMaker.memoryDB().allocateStartSize(20 * 1024 * 1024 - 10000).make();
        StoreWAL wal = (StoreWAL) Store.forDB(db);
        //TODO check WAL size increment
//        assertEquals(1024 * 1024, wal.wal.curVol.length());
        assertEquals(20*1024*1024, wal.vol.length());
        db.close();
    }

    @Test public void allocate_start_size_file(){
        DB db = DBMaker.fileDB(TT.tempDbFile()).allocateStartSize(20 * 1024*1024 -10000).make();
        StoreWAL wal = (StoreWAL) Store.forDB(db);
        assertEquals(20*1024*1024, wal.vol.length());
        db.close();
    }


    @Test public void allocate_start_size_mmap(){
        DB db = DBMaker.fileDB(TT.tempDbFile()).fileMmapEnable().allocateStartSize(20 * 1024*1024 -10000).make();
        StoreWAL wal = (StoreWAL) Store.forDB(db);
        //TODO check WAL size increment
//        assertEquals(1024*1024, wal.wal.curVol.length());
        assertEquals(20*1024*1024, wal.vol.length());
        db.close();
    }


    @Test public void allocate_increment(){
        DB db = DBMaker.memoryDB().allocateIncrement(20 * 1024 * 1024 - 10000).make();
        StoreWAL wal = (StoreWAL) Store.forDB(db);
        //TODO check WAL size increment
//        assertEquals(1024 * 1024, wal.wal.curVol.length());
        assertEquals(32*1024*1024, wal.realVol.length());
        wal.realVol.ensureAvailable(35 * 1024 * 1024);
        assertEquals(64 * 1024 * 1024, wal.realVol.length());

        db.close();
    }


    @Test public void allocate_increment_mmap(){
        DB db = DBMaker.fileDB(TT.tempDbFile()).fileMmapEnable().allocateIncrement(20 * 1024 * 1024 - 10000).make();
        StoreWAL wal = (StoreWAL) Store.forDB(db);
        //TODO check WAL size increment
//        assertEquals(1024 * 1024, wal.wal.curVol.length());
        assertEquals(32*1024*1024, wal.realVol.length());
        wal.realVol.ensureAvailable(35 * 1024 * 1024);
        assertEquals(64 * 1024 * 1024, wal.realVol.length());

        db.close();
    }

    @Test public void serializer_class_loader(){
        final Set<String> loadedClasses = new HashSet<String>();
        ClassLoader l = new ClassLoader() {
            @Override
            public Class<?> loadClass(String name) throws ClassNotFoundException {
                loadedClasses.add(name);
                return super.loadClass(name);
            }
        };
        DB db = DBMaker.memoryDB().serializerClassLoader(l).transactionDisable().make();

        TT.clone(new Class1(), db.getDefaultSerializer());
        assertTrue(loadedClasses.contains(Class1.class.getName()));

        db.close();
        loadedClasses.clear();

        db = DBMaker.memoryDB()
                .serializerRegisterClass(Class2.class.getName(),l)
                .transactionDisable()
                .make();

        TT.clone(new Class2(), db.getDefaultSerializer());
        assertTrue(loadedClasses.contains(Class2.class.getName()));
        db.close();
    }

    public static class Class1 implements Serializable {
    }

    public static class Class2 implements Serializable {
    }

    @Test public void cc() throws IllegalAccessException {
        assertEquals(CC.DEFAULT_CACHE, DBMaker.CC().get("DEFAULT_CACHE"));
    }

    @Test public void fileMmapPreclearDisable1(){
        File f = TT.tempDbFile();
        StoreDirect d = (StoreDirect) DBMaker
                .fileDB(f)
                .fileMmapEnable()
                .fileMmapPreclearDisable()
                .transactionDisable()
                .makeEngine();

        assertTrue(((Volume.MappedFileVol)d.vol).preclearDisabled);
        d.close();
        f.delete();
    }

    @Test public void fileMmapPreclearDisable2(){
        File f = TT.tempDbFile();
        StoreDirect d = (StoreDirect) DBMaker
                .fileDB(f)
                .fileMmapEnable()
                .transactionDisable()
                .makeEngine();

        assertFalse(((Volume.MappedFileVol)d.vol).preclearDisabled);
        d.close();
        f.delete();
    }

}

File: src/test/java/org/mapdb/DBMakerTest.kt
package org.mapdb

import org.junit.Assert.*
import org.junit.Rule
import org.junit.Test
import org.junit.rules.ExpectedException
import org.mapdb.StoreAccess.volume
import org.mapdb.VolumeAccess.sliceShift
import org.mapdb.elsa.Bean1
import org.mapdb.volume.ByteArrayVol
import org.mapdb.volume.FileChannelVol
import org.mapdb.volume.MappedFileVol
import org.mapdb.volume.RandomAccessFileVol
import java.util.*

class DBMakerTest{

    @Rule @JvmField
    val expectedException = ExpectedException.none()!!

    @Test fun sharded_htreemap_close(){
        val executor = TT.executor()

        val map = DBMaker.heapShardedHashMap(8).expireExecutor(executor).expireAfterCreate(100).create()
        assertTrue(executor.isShutdown.not())
        map.close()
        assertTrue(executor.isShutdown)
        assertTrue(executor.isTerminated)
    }

    @Test fun conc_scale(){
        val db =DBMaker.memoryDB().concurrencyScale(32).make()
        assertEquals(DataIO.shift(32), (db.store as StoreDirect).concShift)
    }


    @Test fun conc_disable(){
        var db =DBMaker.memoryDB().make()
        assertTrue(db.isThreadSafe)
        assertTrue(db.store.isThreadSafe)
        assertTrue(db.hashMap("aa1").create().isThreadSafe)
        assertTrue(db.treeMap("aa2").create().isThreadSafe)

        db =DBMaker.memoryDB().concurrencyDisable().make()
        assertFalse(db.isThreadSafe)
        assertFalse(db.store.isThreadSafe)
        assertFalse(db.hashMap("aa1").create().isThreadSafe)
        assertFalse(db.treeMap("aa2").create().isThreadSafe)
    }

    @Test fun raf(){
        val file = TT.tempFile()
        val db = DBMaker.fileDB(file).make()
        assertTrue((db.store as StoreDirect).volumeFactory == RandomAccessFileVol.FACTORY)
        file.delete()
    }

    @Test fun channel(){
        val file = TT.tempFile()
        val db = DBMaker.fileDB(file).fileChannelEnable().make()
        assertTrue((db.store as StoreDirect).volumeFactory == FileChannelVol.FACTORY)
        file.delete()
    }


    @Test fun mmap(){
        val file = TT.tempFile()
        val db = DBMaker.fileDB(file).fileMmapEnable().make()
        assertTrue((db.store as StoreDirect).volumeFactory is MappedFileVol.MappedFileFactory)
        file.delete()
    }


    @Test fun mmap_if_supported(){
        val file = TT.tempFile()
        val db = DBMaker.fileDB(file).fileChannelEnable().fileMmapEnableIfSupported().make()
        if(DataIO.JVMSupportsLargeMappedFiles())
            assertTrue((db.store as StoreDirect).volumeFactory is MappedFileVol.MappedFileFactory)
        else
            assertTrue((db.store as StoreDirect).volumeFactory == FileChannelVol.FACTORY)

        file.delete()
    }


    @Test fun readonly_vol(){
        val f = TT.tempFile()
        //fill with content
        var db = DBMaker.fileDB(f).make()
        db.atomicInteger("aa",1)
        db.close()

        fun checkReadOnly(){
            assertTrue(((db.store) as StoreDirect).volume.isReadOnly)
            TT.assertFailsWith(UnsupportedOperationException::class.java){
                db.hashMap("zz").create()
            }
        }

        db = DBMaker.fileDB(f).readOnly().make()
        checkReadOnly()
        db.close()

        db = DBMaker.fileDB(f).readOnly().fileChannelEnable().make()
        checkReadOnly()
        db.close()

        db = DBMaker.fileDB(f).readOnly().fileMmapEnable().make()
        checkReadOnly()
        db.close()

        f.delete()
    }

    @Test fun checksumStore(){
        val db = DBMaker.memoryDB().checksumStoreEnable().make()
        assertTrue(((db.store) as StoreDirect).checksum)
    }

    @Test(timeout=10000)
    fun file_lock_wait(){
        val f = TT.tempFile()
        val db1 = DBMaker.fileDB(f).make()
        TT.fork{
            Thread.sleep(2000)
            db1.close()
        }
        val db2 = DBMaker.fileDB(f).fileLockWait(6000).make()
        db2.close()
        f.delete()
    }


    @Test(timeout=10000)
    fun file_lock_wait2(){
        val f = TT.tempFile()
        val db1 = DBMaker.fileDB(f).make()
        TT.fork{
            Thread.sleep(2000)
            db1.close()
        }
        val db2 = DBMaker.fileDB(f).fileLockWait().make()
        db2.close()
        f.delete()
    }

    @Test(timeout=10000)
    fun file_lock_wait_time_out_same_jvm() {
        val f = TT.tempFile()

        val db1 = DBMaker.fileDB(f)
                .make()

        try {
            expectedException.expect(DBException.FileLocked::class.java)
            DBMaker.fileDB(f)
                    .fileLockWait(2000)
                    .make()
        } finally {
            db1.close()
            f.delete()
        }
    }

    @Test(timeout=10000)
    fun file_lock_wait_time_out() {
        val f = TT.tempFile()
        val db1 = DBMaker.fileDB(f).make()
        TT.fork {
            Thread.sleep(2000) // Test succeeds if this is commented out
        }

        val maker = DBMaker.fileDB(f).fileLockWait(500)
        try {
            maker.make()
        } catch (e : DBException.FileLocked) {
            // Expected
        }

        db1.close()
        f.delete()
    }

    @Test(timeout=30000)
    fun file_lock_wait_time_out_different_jvm() {
        val f = TT.tempFile()
        val process = TT.forkJvm(ForkedLockTestMain::class.java, f.absolutePath)

        // Wait for the forked process to write to STDOUT, which happens after it
        // has successfully opened and locked the database.
        process.inputStream.read()

        try {
            expectedException.expect(DBException.FileLocked::class.java)
            DBMaker.fileDB(f)
                    .fileLockWait(2000)
                    .make()
        } finally {
            if(!process.isAlive) {
                fail(process.errorStream.reader().readText())
            } else {
                process.destroyForcibly()
                f.delete()
            }
        }
    }

    @Test fun file_lock_disable_RAF(){
        val f = TT.tempFile()
        val db1 = DBMaker.fileDB(f).make()
        DBMaker.fileDB(f).fileLockDisable().make()
    }

    @Test fun file_lock_disable_RAF2(){
        val f = TT.tempFile()
        val db1 = DBMaker.fileDB(f).transactionEnable().make()
        DBMaker.fileDB(f).fileLockDisable().transactionEnable().make()
    }

    @Test fun file_lock_disable_Channel(){
        val f = TT.tempFile()
        val db1 = DBMaker.fileDB(f).make()
        DBMaker.fileDB(f).fileLockDisable().make()
    }

    @Test fun file_lock_disable_Channel2(){
        val f = TT.tempFile()
        val db1 = DBMaker.fileDB(f).fileChannelEnable().transactionEnable().make()
        DBMaker.fileDB(f).fileChannelEnable().fileLockDisable().transactionEnable().make()
    }

    @Test fun file_lock_disable_mmap(){
        val f = TT.tempFile()
        val db1 = DBMaker.fileDB(f).fileMmapEnable().make()
        DBMaker.fileDB(f).fileLockDisable().make()
    }

    @Test fun file_lock_disable_mmap2(){
        val f = TT.tempFile()
        val db1 = DBMaker.fileDB(f).transactionEnable().make()
        DBMaker.fileDB(f).fileLockDisable().fileMmapEnable().transactionEnable().make()
    }

    @Test fun fileIncrement(){
        val db = DBMaker.memoryDB().allocateIncrement(100).make()
        val store = db.store as StoreDirect
        val volume = store.volume as ByteArrayVol
        assertEquals(CC.PAGE_SHIFT, volume.sliceShift)
    }


    @Test fun fileIncrement2(){
        val db = DBMaker.memoryDB().allocateIncrement(2*1024*1024).make()
        val store = db.store as StoreDirect
        val volume = store.volume as ByteArrayVol
        assertEquals(1+CC.PAGE_SHIFT, volume.sliceShift)
    }


    @Test fun fromVolume(){
        val vol = ByteArrayVol()
        val db = DBMaker.volumeDB(vol, false).make()
        assertTrue(vol === (db.store as StoreDirect).volume)
    }



    @Test fun classLoader(){
        val classes = ArrayList<String>()
        val cl = object: ClassLoader() {
            override fun loadClass(name: String?): Class<*> {
                classes+= name!!
                return super.loadClass(name)
            }
        }

        val db = DBMaker.memoryDB().classLoader(cl).make()
        assertEquals(db.classLoader, cl )

        val m = db.atomicVar("a").create()
        m.set(Bean1("aa","bb"))
        m.get()

        assert(classes.contains(Bean1::class.java.name))
    }


    object ForkedLockTestMain {
        @JvmStatic
        fun main(args : Array<String>) {
            if(args.size != 1) {
                System.err.println("No database specified!")
                System.exit(3)
            }

            val file = args[0]
            val db1 = DBMaker.fileDB(file).make()
            System.out.println("Locked database.")
            Thread.sleep(60000)
            db1.close()
        }
    }
}
